File: yukino/mcp/strategy.go
// Copyright (c) 2026 hangtiancheng
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package mcp

import (
	"bytes"
	"encoding/json"
	"net/url"
	"os"
	"strings"

	"github.com/hangtiancheng/yukino-code/yukino/tools"
	"github.com/hangtiancheng/yukino-code/yukino/utils"
)

// Determines how MCP tools enter the context. Three paths, decided once at
// session start after connecting to MCP:
//
//	eager    — total MCP schema size is under 10% of context; load everything
//	           into tools[] with no deferral. The small context savings are not
//	           worth any additional risk.
//	native   — official Anthropic endpoint. Tools stay in tools[] with
//	           defer_loading but the server hides them from the model;
//	           ToolSearch returns tool_reference so the server expands the schema.
//	Endpoints other than native do not support either mechanism and must
//	simulate it: MCP tools never enter tools[] and calls go through McpCall.
//
// Why three paths: tools are rendered after system and before messages; any
// array change invalidates the entire trailing conversation history cache.
// Measured with 20k-token history, appending one tool to tools drops the hit
// rate from 99.4% to 9.5% — effectively recomputing the entire history.
const (
	// DefaultEagerThresholdPercent: below this fraction of the context window, skip deferral
	DefaultEagerThresholdPercent = 10

	// CharsPerToken is the estimation ratio used when real token counts are
	// unavailable. MCP schemas are JSON with high symbol density, so they
	// have fewer characters per token than natural language.
	CharsPerToken = 2.5

	// NativeToolSearchBeta is the beta header for the official endpoint;
	// both defer_loading and tool_reference require it.
	NativeToolSearchBeta = "advanced-tool-use-2025-11-20"

	envLoadingOverride = "YUKINO_MCP_LOADING"
)

var officialHosts = map[string]bool{"api.anthropic.com": true}

// IsOfficialAnthropicEndpoint reports whether the endpoint is official.
// An empty baseURL means the SDK default address, which is official.
// Like the TS `new URL(baseUrl)`, a baseURL without a scheme is invalid
// (scheme-relative "//api.anthropic.com" throws in TS) and reports false.
func IsOfficialAnthropicEndpoint(baseURL string) bool {
	if baseURL == "" {
		return true
	}
	u, err := url.Parse(baseURL)
	if err != nil || u.Scheme == "" {
		return false
	}
	return officialHosts[strings.ToLower(u.Hostname())]
}

// EstimateSchemaTokens estimates token count from character count.
func EstimateSchemaTokens(schemaChars int) int {
	return int(float64(schemaChars) / CharsPerToken)
}

// DecideMode determines the loading mode.
func DecideMode(baseURL string, contextWindow, mcpSchemaChars, thresholdPercent int) tools.McpLoadingMode {
	switch strings.ToLower(strings.TrimSpace(os.Getenv(envLoadingOverride))) {
	case "eager":
		return tools.McpLoadingEager
	case "native":
		return tools.McpLoadingNative
	case "dispatch":
		return tools.McpLoadingDispatch
	}

	// No MCP tools; any path behaves the same, eager is simplest
	if mcpSchemaChars <= 0 {
		return tools.McpLoadingEager
	}

	budget := float64(contextWindow) * float64(thresholdPercent) / 100
	if float64(EstimateSchemaTokens(mcpSchemaChars)) < budget {
		return tools.McpLoadingEager
	}
	if IsOfficialAnthropicEndpoint(baseURL) {
		return tools.McpLoadingNative
	}
	return tools.McpLoadingDispatch
}

// MeasureSchemaChars counts the serialized character length of MCP tool
// schemas, used for threshold comparison. TS measures
// JSON.stringify(schema).length — UTF-16 code units of unescaped JSON
// (JSON.stringify does not HTML-escape), so a CJK description counts one
// unit per character, not three bytes.
func MeasureSchemaChars(registry *tools.Registry) int {
	total := 0
	for _, t := range registry.ListTools() {
		if !strings.HasPrefix(t.Name(), tools.MCPToolPrefix) {
			continue
		}
		if b, err := marshalUnescaped(t.Schema()); err == nil {
			total += utils.UTF16Len(string(b))
		} else {
			total += utils.UTF16Len(t.Name()) + utils.UTF16Len(t.Description())
		}
	}
	return total
}

// marshalUnescaped serializes v like JSON.stringify: standard JSON with
// SetEscapeHTML(false), since encoding/json would otherwise expand
// <, > and & into \u003c-style sequences TS never emits.
func marshalUnescaped(v any) ([]byte, error) {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(v); err != nil {
		return nil, err
	}
	// Encode appends a newline; JSON.stringify has none.
	return bytes.TrimRight(buf.Bytes(), "\n"), nil
}

// ApplyMode applies the decision to the registry.
//
// Under eager, the defer flag on MCP tools is cleared so they appear in
// tools[]; the other two paths keep them deferred. McpCall is not registered
// here — it must already be in tools[] before the MCP connection is
// established; adding it after connection would be a mid-session tools array
// change, breaking the cache just the same.
func ApplyMode(registry *tools.Registry, mode tools.McpLoadingMode) {
	registry.McpLoadingMode = mode
	eager := mode == tools.McpLoadingEager
	for _, t := range registry.ListTools() {
		if mt, ok := t.(tools.MCPTool); ok {
			mt.SetDeferLoading(!eager)
		}
	}

	// Search and dispatch exposure is decided by mode. Under eager all tools
	// are in tools[]; there is nothing to search and no dispatch entry point
	// needed. These two flags are computed once here and fixed for the entire
	// session, causing no mid-session tools[] churn.
	registry.ExposeToolSearch = !eager
	registry.ExposeMcpCall = mode == tools.McpLoadingDispatch
}

// DecideAndApply is the single entry point called after connecting to MCP.
func DecideAndApply(registry *tools.Registry, baseURL string, contextWindow int) tools.McpLoadingMode {
	mode := DecideMode(baseURL, contextWindow, MeasureSchemaChars(registry), DefaultEagerThresholdPercent)
	ApplyMode(registry, mode)
	return mode
}

File: apps/yukino/src/mcp/strategy.ts
/**
 * Copyright (c) 2026 hangtiancheng
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * Decides how MCP tools enter the context. Three modes, chosen once at session start
 * right after MCP connects:
 *
 *   eager    Total schema size is under one tenth of the context window, so load
 *            everything into tools[] with no deferral. The context saved is not worth
 *            taking on any extra risk for.
 *   native   Official Anthropic endpoints. Tools stay in tools[] with defer_loading but
 *            the server hides them from the model; ToolSearch returns tool_reference and
 *            the server expands the schema.
 *   dispatch Other endpoints (domestic vendors, various proxy gateways) support neither
 *            of the above, so we simulate it ourselves: MCP tools never enter tools[] and
 *            everything goes through the single McpCall entry point.
 *
 * Why three modes: tools is rendered after system and before messages, so any change to
 * the array invalidates the prompt cache for the entire conversation history that follows.
 * Measured with a 20k-token history, appending one tool to the end of tools drops the
 * cache hit rate from 99.4% to 9.5%, effectively recomputing the whole history.
 */

import { MCP_TOOL_PREFIX } from "./tool-wrapper.js";

import { isMcpToolLike } from "@/tools/mcp-call.js";
import type { ToolRegistry } from "@/tools/registry.js";
import type { McpLoadingMode } from "@/tools/types.js";

/** Below this share of the context window, skip deferral and load everything eagerly. */
export const DEFAULT_EAGER_THRESHOLD_PERCENT = 10;

/**
 * Estimation ratio used when real token counts are unavailable. MCP schemas are JSON
 * with high symbol density, so they have fewer characters per token than natural language.
 */
export const CHARS_PER_TOKEN = 2.5;

/** Beta header for official endpoints; defer_loading and tool_reference both require it. */
export const NATIVE_TOOL_USE_BETA = "advanced-tool-use-2025-11-20";

const OFFICIAL_HOSTS = new Set(["api.anthropic.com"]);
const ENV_OVERRIDE = "YUKINO_MCP_LOADING";

/** An empty baseUrl means the SDK default address, i.e. the official one. */
export function isOfficialAnthropicEndpoint(baseUrl: string): boolean {
  if (!baseUrl) {
    return true;
  }
  try {
    return OFFICIAL_HOSTS.has(new URL(baseUrl).hostname.toLowerCase());
  } catch {
    return false;
  }
}

export function estimateSchemaTokens(schemaChars: number): number {
  return Math.floor(schemaChars / CHARS_PER_TOKEN);
}

export function decideMode(
  baseUrl: string,
  contextWindow: number,
  mcpSchemaChars: number,
  thresholdPercent = DEFAULT_EAGER_THRESHOLD_PERCENT,
): McpLoadingMode {
  const override = (process.env[ENV_OVERRIDE] ?? "").trim().toLowerCase();
  if (
    override === "eager" ||
    override === "native" ||
    override === "dispatch"
  ) {
    return override;
  }

  // No MCP tools, any mode would behave the same, and eager is the simplest
  if (mcpSchemaChars <= 0) {
    return "eager";
  }

  const budget = (contextWindow * thresholdPercent) / 100;
  if (estimateSchemaTokens(mcpSchemaChars) < budget) {
    return "eager";
  }

  return isOfficialAnthropicEndpoint(baseUrl) ? "native" : "dispatch";
}

/** Character count of the serialized MCP tool schemas, for comparing against the threshold. */
export function measureSchemaChars(registry: ToolRegistry): number {
  let total = 0;
  for (const tool of registry.listTools()) {
    if (!tool.name.startsWith(MCP_TOOL_PREFIX)) {
      continue;
    }
    try {
      total += JSON.stringify(tool.schema()).length;
    } catch {
      total += tool.name.length + (tool.description?.length ?? 0);
    }
  }
  return total;
}

/**
 * Applies the decision to the registry.
 *
 * Under eager, the defer flag on MCP tools must be cleared so they appear in tools[];
 * the other two modes keep them deferred. McpCall is not registered here — it must
 * already be in tools[] before the MCP connection is established, otherwise adding it
 * after connecting would be a mid-session tools[] mutation and break the cache just the same.
 */
export function applyMode(registry: ToolRegistry, mode: McpLoadingMode): void {
  registry.mcpLoadingMode = mode;
  const eager = mode === "eager";
  for (const tool of registry.listTools()) {
    if (isMcpToolLike(tool) && typeof tool.setDeferLoading === "function") {
      tool.setDeferLoading(!eager);
    }
  }

  // Search and dispatch exposure is decided per mode. Under eager every tool is already in
  // tools[], so there is nothing to search and no need for a dispatch entry point. These two
  // flags are computed once here and fixed for the whole session, so tools[] never fluctuates.
  registry.exposeToolSearch = !eager;
  registry.exposeMcpCall = mode === "dispatch";
}

/** Entry point called once after MCP connects. */
export function decideAndApply(
  registry: ToolRegistry,
  baseUrl: string,
  contextWindow: number,
): McpLoadingMode {
  const mode = decideMode(baseUrl, contextWindow, measureSchemaChars(registry));
  applyMode(registry, mode);
  return mode;
}
